neonvm/controllers/virtualmachine_qmp_queries.go

```diff
@@ -229,72 +245,380 @@ searchForEmpty:
+				fmt.Sprintf("Failed to add memslot%d to VM %s: %s",
+					idx, r.vm.Name, err.Error()))
+			continue
+		}
+		r.recorder.Event(r.vm, "Normal", "ScaleUp",
+			fmt.Sprintf("Added memslot%d to VM %s", idx, r.vm.Name))
+		r.memBackends[idx] = false
+		r.maxBackend = max(r.maxBackend, idx)
 	}
+	return attempts.DidSomething()
+}
 
-	// add memdev object
-	cmd := []byte(fmt.Sprintf(`{"execute": "object-add", "arguments": {"id": "memslot%d", "size": %d, "qom-type": "memory-backend-ram"}}`, slot, virtualmachine.Spec.Guest.MemorySlotSize.Value()))
-	_, err = mon.Run(cmd)
+func (r *QmpMemorySetter) AddDevices() bool {
+	attempts := NewAttemptsCounter(r.targetCnt - r.memDevCount)
+	for idx := 1; idx <= r.maxBackend; idx++ {
+		active, ok := r.memBackends[idx]
+		if !ok || active {
+			continue
+		}
+		// Found unused backend to plug into
+
+		if !attempts.Attempt() {
+			break
+		}
+
+		err := QmpAddMemoryDevice(r.mon, idx)
+		if err != nil {
+			r.errs = append(r.errs, err)
+			r.recorder.Event(r.vm, "Warning", "ScaleUp",
+				fmt.Sprintf("Failed to add dimm%d to VM %s: %s",
+					idx, r.vm.Name, err.Error()))
+			continue
+		}
+		r.recorder.Event(r.vm, "Normal", "ScaleUp",
+			fmt.Sprintf("Added dimm%d to VM %s", idx, r.vm.Name))
+		r.memBackends[idx] = true
+		r.memDevCount++
+	}
+	return attempts.DidSomething()
+}
+
+func (r *QmpMemorySetter) RemoveDevices() bool {
+	attempts := NewAttemptsCounter(r.memDevCount - r.targetCnt)
+	// Removing from the end to keep memslot1,memslot2,...
+	for idx := r.maxBackend; idx >= 1; idx-- {
+		active, ok := r.memBackends[idx]
+		if !ok || !active {
+			continue
+		}
+		// Found used backend to remove
+
+		if !attempts.Attempt() {
+			break
+		}
+
+		err := QmpDelMemoryDevice(r.mon, idx)
+		if err != nil {
+			r.errs = append(r.errs, err)
+			r.recorder.Event(r.vm, "Warning", "ScaleDown",
+				fmt.Sprintf("Failed to remove dimm%d from VM %s: %s",
+					idx, r.vm.Name, err.Error()))
+			continue
+		}
+		r.recorder.Event(r.vm, "Normal", "ScaleDown",
+			fmt.Sprintf("Removed dimm%d from VM %s", idx, r.vm.Name))
+		r.memBackends[idx] = false
+		r.memDevCount--
+	}
+
+	return attempts.DidSomething()
+}
+
+func (r *QmpMemorySetter) RemoveBackends() bool {
+	attempts := NewAttemptsCounter(len(r.memBackends) - r.targetCnt)
+
+	for idx := r.maxBackend; idx >= 1; idx-- {
+		active, ok := r.memBackends[idx]
+		if !ok || active {
+			continue
+		}
+
+		if !attempts.Attempt() {
+			break
+		}
+
+		err := QmpDelMemoryBackend(r.mon, idx)
+		if err != nil {
+			r.errs = append(r.errs, err)
+			r.recorder.Event(r.vm, "Warning", "ScaleDown",
+				fmt.Sprintf("Failed to remove memslot%d from VM %s: %s",
+					idx, r.vm.Name, err.Error()))
+			continue
+		}
+		r.recorder.Event(r.vm, "Normal", "ScaleDown",
+			fmt.Sprintf("Removed memslot%d from VM %s", idx, r.vm.Name))
+		delete(r.memBackends, idx)
+	}
+	return attempts.DidSomething()
+}
+
+func (r *QmpMemorySetter) run() (int, error) {
+	fmt.Println("start run")
```

Comment:
probably you want to remove this or keep it with proper logging

Reply:
Removed, thx